image_store/backends.py

```diff
@@ -0,0 +1,215 @@
+
+            # create course space if not exists for storing images
+            course = self._find_or_create_course()
+
+            # upload image to the course space
+            course_images = self._upload_images(course['id'], uploaded_files)
+            course_image_ids = [image['id'] for image in course_images]
+            
+            # create a collection and add images to it
+            collection = self._create_collection(course['id'], title)
+            self._add_to_collection(collection['id'], course_image_ids)
+            
+            # get IIIF manifest URL for collection
+            collection = self._get_collection(collection['id'])
+            manifest_url = collection.get('iiif_manifest', {}).get('url')
+
+        except ImageStoreBackendException as e:
+            logger.exception("Image store backend error: %s" % str(e))
+            raise e
+
+        return manifest_url
+
+    def _authenticate(self):
+        ''' 
+        Authenticates with API using client credentials (key/secret).
+        Obtains temporary access token for subsequent API requests.
+        '''
+        url = "%s/auth/obtain-token" % self.base_url
+        post_data = {
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+            "user_id": self.user_id,
+        }
+        r = requests.post(url, headers=self.headers, data=json.dumps(post_data))
+        logger.debug("API POST {url} data:{post_data} status:{status_code} text:{text}".format(url=url, post_data=post_data, status_code=r.status_code, text=r.text))
+        
+        access_token = None
+        if r.status_code == 200:
+            data = r.json()
+            access_token = data.get('access_token', None)
+            if not access_token:
+                raise ImageStoreBackendException("Authenticated successfully, but access token absent from response")
+        else: 
+            raise ImageStoreBackendException("Authentication failed: {status_code}".format(status_code=r.status_code))
+        return access_token
+    
+    def _find_or_create_course(self):
+        '''
+        Find associated course space for images using Canvas/LTI identifiers.
+        If no course space exists, create one.
+        '''
+        url = "%s/courses" % self.base_url
+        search_params = {
+            "lti_context_id": self.course_attrs["lti_context_id"],
+            "lti_tool_consumer_instance_guid": self.course_attrs["lti_tool_consumer_instance_guid"]
+        }
+        r = requests.get(url, headers=self.headers, params=search_params)
+        logger.debug("API GET {url} params:{params} status:{status_code} text:{text}".format(url=url, params=search_params, status_code=r.status_code, text=r.text))
+
+        course = None
+        if r.status_code == 200:
+            courses = r.json()
+            if len(courses) == 0:
+                course = self._create_course()
+            else:
+                course = courses[0]
```

Comment:
Is it ever possible that the `course` object that we want wont be located at `[0]` of the json response?

Reply:
Good question. No I would expect a `200` response to only ever return 0 or 1 courses. The reason being is that our search parameters match a unique constraint on the course API model:

```python
unique_together = ("lti_context_id", "lti_tool_consumer_instance_guid")
```

Given that's the case, I think I'll change this to throw an exception if more than one courses are returned, since that shouldn't happen and would indicate a problem.